server/src/main/java/org/opensearch/telemetry/tracing/listener/TraceableSearchRequestOperationsListener.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.telemetry.tracing.listener;
+
+import org.opensearch.action.search.SearchPhaseContext;
+import org.opensearch.action.search.SearchRequestContext;
+import org.opensearch.action.search.SearchRequestOperationsListener;
+import org.opensearch.telemetry.tracing.AttributeNames;
+import org.opensearch.telemetry.tracing.Span;
+import org.opensearch.telemetry.tracing.SpanBuilder;
+import org.opensearch.telemetry.tracing.SpanContext;
+import org.opensearch.telemetry.tracing.SpanScope;
+import org.opensearch.telemetry.tracing.Tracer;
+
+import static org.opensearch.core.common.Strings.capitalize;
+
+/**
+ * SearchRequestOperationsListener subscriber for search request tracing
+ *
+ * @opensearch.internal
+ */
+public final class TraceableSearchRequestOperationsListener extends SearchRequestOperationsListener {
+    private final Tracer tracer;
+    private final Span requestSpan;
+    private Span phaseSpan;
+    private SpanScope phaseSpanScope;
+
+    public TraceableSearchRequestOperationsListener(final Tracer tracer, final Span requestSpan) {
+        this.tracer = tracer;
+        this.requestSpan = requestSpan;
+        this.phaseSpan = null;
+        this.phaseSpanScope = null;
+
+        if (tracer.isRecording() == false) {
+            this.setEnabled(false);
```

Comment:
For my understanding please explain, how this enabled flag is being used. We should be skipping this listener altogether if tracing is disabled.

Reply:
We create an instance of `TraceableSearchRequestOperationsListener` regardless of whether tracing is on or off, but set the `enabled` flag accordingly. Then when the composite listener is created with buildCompositeListener() method pasted below, `enabled` flag is checked - see `.filter(..)` line. Here the `TraceableSearchRequestOperationsListener` will be filtered **out** of `searchListenersList` if tracing is off and cannot be executed.

```
    public SearchRequestOperationsListener.CompositeListener buildCompositeListener(
        final SearchRequest searchRequest,
        final Logger logger,
        final SearchRequestOperationsListener... perRequestListeners
    ) {
        final List<SearchRequestOperationsListener> searchListenersList = Stream.concat(
            searchRequestListenersList.stream(),
            Arrays.stream(perRequestListeners)
        )
            .filter((searchRequestOperationsListener -> searchRequestOperationsListener.isEnabled(searchRequest)))
            .collect(Collectors.toList());

        return new SearchRequestOperationsListener.CompositeListener(searchListenersList, logger);
    }
```

Reply:
@dzane17 please follow the `TraceableActionListener` (& others) pattern:

```
    public static SearchRequestOperationsListener(final Tracer tracer, final Span requestSpan) { create(Span span, Tracer tracer) {
        if (tracer.isRecording() == true) {
            return new TraceableSearchRequestOperationsListener(...);
        } else {
            return NOOP;
        }
    }

```

The `NOOP` could be a static final (single instance), so it won't be created all the time